File: Practica16 Deteccion de Rostros/Matlab/faceDetection.m
% Crear un objeto de video para la webcam
cam = webcam(3);
cam.Resolution = '640x480';

% Crear un visor de video
videoPlayer = vision.VideoPlayer('Position', [100, 100, 600, 400]);

% Crear un detector de rostros
faceDetector = vision.CascadeObjectDetector();

% Ejecutar el bucle para la detección de rostros en tiempo real
while true
    % Capturar un fotograma de la webcam
    frame = snapshot(cam);

    % Convertir la imagen de RGB a escala de grises
    grayFrame = rgb2gray(frame);

    % Detectar rostros en la imagen en escala de grises
    bbox = step(faceDetector, grayFrame);

    % Dibujar rectángulos alrededor de los rostros detectados
    detectedImg = insertObjectAnnotation(frame, 'rectangle', bbox, 'Face');

    % Mostrar la imagen con los rostros detectados
    step(videoPlayer, detectedImg);

    if ~isOpen(videoPlayer)
        break;
    end

end

% Liberar recursos
release(videoPlayer);
clear cam;

File: Practica16 Deteccion de Rostros/Python/faceDetection.py
import cv2

# Cargar el clasificador de detección de rostros preentrenado
# Para ver más modelos preentrenados:
# https://github.com/opencv/opencv/tree/master/data/haarcascades

face_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + 'haarcascade_frontalface_default.xml')
# face_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + 'haarcascade_profileface.xml')
# face_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + 'haarcascade_eye.xml')

# Iniciar la webcam
cap = cv2.VideoCapture(2)

while True:
    # Capturar fotograma por fotograma
    ret, frame = cap.read()
    if not ret:
        break

    # Convertir el fotograma a escala de grises
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

    # Detectar rostros en el fotograma
    faces = face_cascade.detectMultiScale(gray, scaleFactor=1.3, minNeighbors=5, minSize=(30, 30))

    # Dibujar rectángulos alrededor de los rostros detectados
    for (x, y, w, h) in faces:
        cv2.rectangle(frame, (x, y), (x+w, y+h), (255, 0, 0), 2)

    # Mostrar el fotograma con los rostros detectados
    cv2.imshow('Face Detection', frame)

    # Salir del bucle si se presiona 'q'
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

# Liberar la webcam y cerrar las ventanas
cap.release()
cv2.destroyAllWindows()
